### PR TITLE
Add a test script and fix bugs

### PR DESCRIPTION
**Pull request by nicelhc13**

### CODE REVIEW

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 1]
```diff
@@ -42,3 +42,125 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+  std::sort(nodes.begin(), nodes.end(), [&graph](GNode& l_opr, GNode& r_opr) {
+    MetisNode& l_data               = graph.getData(l_opr);
+    MetisNode& r_data               = graph.getData(r_opr);
+    float l_gain                    = l_data.getGain();
+    float r_gain                    = r_data.getGain();
+    float l_weight                  = l_data.getWeight();
+    float r_weight                  = r_data.getWeight();
+    float l_cost                    = l_gain / l_weight;
+    float r_cost                    = r_gain / r_weight;
+    [[maybe_unused]] uint32_t l_nid = l_data.nodeid;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:56]: remove these if not used.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 2]
```diff
@@ -42,3 +42,125 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+
+void InitGain(GGraph& g) {
+  galois::do_all(
+      galois::iterate(g.hedges, g.size()),
+      [&](GNode n) {
+        MetisNode& node    = g.getData(n);
+        node.positive_gain = 0;
+        node.negative_gain = 0;
+      },
+      galois::loopname("firstinit"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:75]: firstinit --> First-init

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 3]
```diff
@@ -42,3 +42,125 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+              galois::atomicAdd(node_data.positive_gain, 1);
+            }
+            if (nodep == (num_p0_nodes + num_p1_nodes)) {
+              // it means that one of p1 or p2 is zero.
+              galois::atomicAdd(node_data.negative_gain, 1);
+            }
+          }
+        }
+      },
+      galois::steal(), galois::loopname("initGainsPart"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:108]: I was following some standard naming convention for these names: Init-Gains

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 4]
```diff
@@ -42,3 +42,125 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+      galois::iterate(uint32_t{0}, total_nodes),
+      [&](GNode n) {
+        auto& node_index_pair   = combined_nodelist[n];
+        GNode node_id           = node_index_pair.first;
+        uint32_t index          = node_index_pair.second;
+        MetisNode& node_data    = g[index]->getData(node_id);
+        node_data.positive_gain = 0;
+        node_data.negative_gain = 0;
+      },
+      galois::loopname("firstinit"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:127]: ditto.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 5]
```diff
@@ -42,3 +42,125 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+              galois::atomicAdd(dst_data.positive_gain, 1);
+            }
+            if (nodep == (num_p0_nodes + num_p1_nodes)) {
+              // it means that one of p1 or p2 is zero.
+              galois::atomicAdd(dst_data.negative_gain, 1);
+            }
+          }
+        }
+      },
+      galois::steal(), galois::loopname("initGainsPart"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:165]: Ditto.

File: lonestar/eda/cpu/bipart-nested/Partitioning.cpp [hunk 6]
```diff
@@ -255,16 +176,32 @@ void PartitionCoarsestGraphs(std::vector<MetisGraph*>& metis_graphs,
     uint32_t node_bag_size = std::distance(node_bag.begin(), node_bag.end());
     std::vector<GNode> node_vec(node_bag_size);
     uint32_t idx{0};
+    uint32_t gain =
+        (process_zero_part) ? first_part_weights : zero_part_weights;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Partitioning.cpp:180]: change part to partition everywhere.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp (hunk 1 above)
Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:56]: No, they are being used. hochan, please remove maybe-unused thing. This creates amiguity.

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 7]
```diff
@@ -17,715 +17,427 @@
...
-            } else {
-              int d = gain * 10.0f - 1;
-              idx   = 10 - d;
-            }
-            for (auto x : b) {
-              nodeListz[idx].push_back(x);
-            }
+          SortNodesByGainAndWeight(g, cand_nodes_vec_arr[idx]);
+        },
+        galois::steal(), galois::loopname("sort bucket"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:287]: Sort-bucket

File: lonestar/eda/cpu/bipart-nested/Helper.cpp (hunk 1 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp]: No, if I remove [[maybe_unused]] then compiler complains and denies to compile it. Let me move those to inside of the if statement.

Comment (gurbinder533) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp]: I am saying just remove those variables. Are they used anywhere?

**Pull request merged**